### ISSUE TITLE
Client Assertion fails with  jti was already used error on first use.

### ISSUE THREAD

**Issue by alexlovelltroy** [authelia/authelia#7790]
### Version

v4.38.10

### Deployment Method

Docker

### Reverse Proxy

Traefik

### Reverse Proxy Version

2.11.7

### Description

When attempting to use pushed authentication requests for oidc clients, using `token_endpoint_auth_method: private_key_jwt` with a registered public key for validating the incoming jwt client assertion, every client assertion is refused with an error that includes a suspicious sql error.

```
Pushed Authorization Request failed with error: Client authentication failed (e.g., unknown client, no client authentication included, or unsupported authentication method). The requested OAuth 2.0 Client could not be authenticated. The jti was already used. Claim 'jti' from 'client_assertion' MUST only be used once. error selecting oauth2 blacklisted JTI with signature '': sql: no rows in result set
```

My suspicion is that `LoadOAuth2PARContext` from `sql_provider.go` is creating it's own error rather than raising the `sql.ErrNoRows` which is expected by `ClientAssertionJWTValid`  in `store.go` which ultimately leads to all jtis being rejected.

## sql_provider.go reference

https://github.com/authelia/authelia/blob/95e3ea73be1e92823a93da7b1c00b96e938ed953/internal/storage/sql_provider.go#L1250-L1264

## store.go reference

https://github.com/authelia/authelia/blob/95e3ea73be1e92823a93da7b1c00b96e938ed953/internal/oidc/store.go#L117-L134

## client_authentication_strategy.go reference

https://github.com/authelia/oauth2-provider/blob/acbadb6280d549dc7cd67c5a50a0dc54b15e94a8/client_authentication_strategy.go#L243-L277

### Reproduction

I have reproduced this several times using the [local example](https://github.com/authelia/authelia/tree/95e3ea73be1e92823a93da7b1c00b96e938ed953/examples/compose/local) with the following sanitized client configuration.

My client application is python code that generates a unique RSA keypair per client and a unique uuid-based jti per request.  The code is not publishable at this time, but I have verified that the client_assertion is correctly formed and signed by using https://jwt.io.

```yaml

identity_providers:
  oidc:
    clients:
      - client_id: 'example-two'
        public: false
        require_pkce: true
        pkce_challenge_method: 'S256'
        redirect_uris:
          - 'https://federation.example.com/callback'
        scopes:
          - 'offline_access'
          - 'offline'
          - 'authelia.bearer.authz'
        audience:
          - 'https://app1.example.com'
          - 'https://app2.example.com'
        grant_types:
          - 'authorization_code'
          - 'refresh_token'
        response_types:
          - 'code'
        response_modes:
          - 'form_post'
        consent_mode: 'explicit'
        require_pushed_authorization_requests: true
        # https://www.authelia.com/configuration/identity-providers/openid-connect/clients/#token_endpoint_auth_method
        # client_secret_basic, client_secret_post, client_secret_jwt, private_key_jwt, or none
        token_endpoint_auth_method: 'private_key_jwt' 
        token_endpoint_auth_signing_alg: 'RS256'
        access_token_signed_response_key_id: 'example-key'
        jwks:
          - key_id: 'example-client-1'
            algorithm: 'RS256'
            use: 'sig'
            key: |
              -----BEGIN PUBLIC KEY-----
               <SNIP>
              -----END PUBLIC KEY-----

```

### Build Information

```shell
Last Tag: v4.38.10
State: tagged clean
Branch: v4.38.10
Commit: b7d5e3ff29917bd0842c8509b4a8338d19f2934e
Build Number: 31324
Build OS: linux
Build Arch: arm64
Build Compiler: gc
Build Date: Sun, 04 Aug 2024 22:15:26 +1000
Extra:

Go:
    Version: go1.22.2
    Module Path: github.com/authelia/authelia/v4
    Executable Path: github.com/authelia/authelia/v4/cmd/authelia
```

### Logs (Authelia)

```shell
2024-08-30 14:26:09 time="2024-08-31T04:26:09+10:00" level=error msg="Pushed Authorization Request failed with error: Client authentication failed (e.g., unknown client, no client authentication included, or unsupported authentication method). The requested OAuth 2.0 Client could not be authenticated. The jti was already used. Claim 'jti' from 'client_assertion' MUST only be used once. error selecting oauth2 blacklisted JTI with signature '': sql: no rows in result set" method=POST path=/api/oidc/pushed-authorization-request remote_ip=192.168.65.1 stack="github.com/authelia/authelia/v4/internal/handlers/handler_oidc_authorization.go:176          OpenIDConnectPushedAuthorizationRequest\ngithub.com/authelia/authelia/v4/internal/middlewares/http_to_authelia_handler_adaptor.go:114 handleRouter.NewHTTPToAutheliaHandlerAdaptor.func23\ngithub.com/authelia/authelia/v4/internal/middlewares/bridge.go:54                            handleRouter.(*BridgeBuilder).Build.func16.1\ngithub.com/authelia/authelia/v4/internal/middlewares/headers.go:91                           SecurityHeadersNoStore.func1\ngithub.com/authelia/authelia/v4/internal/middlewares/headers.go:75                           SecurityHeadersCSPNoneOpenIDConnect.func1\ngithub.com/authelia/authelia/v4/internal/middlewares/headers.go:30                           SecurityHeadersBase.func1\ngithub.com/authelia/authelia/v4/internal/middlewares/cors.go:216                             handleRouter.(*CORSPolicy).Middleware.func24\ngithub.com/fasthttp/router@v1.5.2/router.go:441                                              (*Router).Handler\ngithub.com/authelia/authelia/v4/internal/middlewares/log_request.go:14                       handleRouter.LogRequest.func40\ngithub.com/authelia/authelia/v4/internal/middlewares/errors.go:38                            RecoverPanic.func1\ngithub.com/valyala/fasthttp@v1.55.0/server.go:2379                                           (*Server).serveConn\ngithub.com/valyala/fasthttp@v1.55.0/workerpool.go:224                                        (*workerPool).workerFunc\ngithub.com/valyala/fasthttp@v1.55.0/workerpool.go:196                                        (*workerPool).getCh.func1\nruntime/asm_arm64.s:1222
```

### Pre-Submission Checklist

- [X] I agree to follow the [Code of Conduct](http://www.authelia.com/code-of-conduct)

- [X] This is a bug report and not a support request

- [X] I have read the security policy and this bug report is not a security issue or security related issue

- [X] I have either included the complete configuration file or I am sure it's unrelated to the configuration

- [X] I have either included the complete debug / trace logs or the output of the build-info command if the logs are not relevant

- [X] I have provided all of the required information in full with the only alteration being reasonable sanitization in accordance with the [Troubleshooting Sanitization](https://www.authelia.com/r/sanitize) reference guide

- [X] I have checked for related proxy or application logs and included them if available

- [X] I have checked for related issues and checked the documentation

**Comment by james-d-elliott** [comment]
Thanks for the detailed report, pretty sure this is accurate without even looking. but I'll add a test to confirm and then fix it.

**Comment by alexlovelltroy** [comment]
It looks like a trivial fix, but I haven't had time yet to rebuild and test. Here's the branch I've started which includes a test.

https://github.com/authelia/authelia/compare/master...alexlovelltroy:authelia:master

**Comment by james-d-elliott** [comment]
Yeah possibly, it's very strange.. it should be functional as is and this passes as I'd expect by the docs regarding errors.Is():

```go
func TestError(t *testing.T) {
	err := fmt.Errorf("error selecting oauth2 blacklisted JTI with signature '%s': %w", "abc", sql.ErrNoRows)

	assert.True(t, errors.Is(err, sql.ErrNoRows))
}
```

**Comment by alexlovelltroy** [comment]
I saw that.  It was part of the reason I started with a bug report rather than just a PR

**Comment by james-d-elliott** [comment]
Does your test fail without the fix?

**Comment by james-d-elliott** [comment]
Looks like it passes regardless. How odd.

**Comment by alexlovelltroy** [comment]
You're right.  In reviewing it again, I'm more confused than before.  Why is it reporting that the signature is empty?  I may have identified the wrong place for this bug.

**Comment by james-d-elliott** [comment]
The signature is empty due to another issue, though that one is inconsequential other than the message itself. See the following patch which resolves that part of the issue:

```diff
Subject: [PATCH] docs: use callouts

Signed-off-by: James Elliott <james-d-elliott@users.noreply.github.com>
---
Index: internal/storage/sql_provider.go
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/internal/storage/sql_provider.go b/internal/storage/sql_provider.go
--- a/internal/storage/sql_provider.go	(revision efc9b462b0e5305f7f64a9f2d2071f2e3266db65)
+++ b/internal/storage/sql_provider.go	(date 1725109896439)
@@ -1304,7 +1304,7 @@
 	blacklistedJTI = &model.OAuth2BlacklistedJTI{}
 
 	if err = p.db.GetContext(ctx, blacklistedJTI, p.sqlSelectOAuth2BlacklistedJTI, signature); err != nil {
-		return nil, fmt.Errorf("error selecting oauth2 blacklisted JTI with signature '%s': %w", blacklistedJTI.Signature, err)
+		return nil, fmt.Errorf("error selecting oauth2 blacklisted JTI with signature '%s': %w", signature, err)
 	}
 
 	return blacklistedJTI, nil
```

**Comment by james-d-elliott** [comment]
This is the full patch I'm working with:

```diff
Subject: [PATCH] docs: use callouts

Signed-off-by: James Elliott <james-d-elliott@users.noreply.github.com>
---
Index: internal/storage/sql_provider.go
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/internal/storage/sql_provider.go b/internal/storage/sql_provider.go
--- a/internal/storage/sql_provider.go	(revision efc9b462b0e5305f7f64a9f2d2071f2e3266db65)
+++ b/internal/storage/sql_provider.go	(date 1725109896439)
@@ -1304,7 +1304,7 @@
 	blacklistedJTI = &model.OAuth2BlacklistedJTI{}
 
 	if err = p.db.GetContext(ctx, blacklistedJTI, p.sqlSelectOAuth2BlacklistedJTI, signature); err != nil {
-		return nil, fmt.Errorf("error selecting oauth2 blacklisted JTI with signature '%s': %w", blacklistedJTI.Signature, err)
+		return nil, fmt.Errorf("error selecting oauth2 blacklisted JTI with signature '%s': %w", signature, err)
 	}
 
 	return blacklistedJTI, nil
Index: internal/oidc/store_test.go
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/internal/oidc/store_test.go b/internal/oidc/store_test.go
--- a/internal/oidc/store_test.go	(revision efc9b462b0e5305f7f64a9f2d2071f2e3266db65)
+++ b/internal/oidc/store_test.go	(date 1725110150166)
@@ -2,9 +2,12 @@
 
 import (
 	"context"
+	"crypto/sha256"
 	"database/sql"
 	"encoding/json"
+	"errors"
 	"fmt"
+	"path"
 	"testing"
 	"time"
 
@@ -638,3 +641,33 @@
 	s.NoError(s.store.MarkJWTUsedForTime(s.ctx, "65471ccb-d650-4006-a95f-cb4f4e3d7202", time.Unix(160000000, 0)))
 	s.EqualError(s.store.MarkJWTUsedForTime(s.ctx, "65471ccb-d650-4006-a95f-cb4f4e3d7201", time.Unix(160000000, 0)), "already marked")
 }
+
+func TestStoreSQLite(t *testing.T) {
+	dir := t.TempDir()
+
+	db := storage.NewSQLiteProvider(&schema.Configuration{
+		Storage: schema.Storage{
+			Local: &schema.StorageLocal{
+				Path: path.Join(dir, "db.sqlite3"),
+			},
+		},
+	})
+
+	err := db.StartupCheck()
+
+	require.NoError(t, err)
+
+	store := oidc.NewStore(&schema.IdentityProvidersOpenIDConnect{}, db)
+
+	jti := "example"
+
+	signature := fmt.Sprintf("%x", sha256.Sum256([]byte(jti)))
+
+	assert.NoError(t, store.ClientAssertionJWTValid(context.Background(), "example"))
+
+	bjti, err := db.LoadOAuth2BlacklistedJTI(context.Background(), signature)
+
+	assert.True(t, errors.Is(err, sql.ErrNoRows))
+	assert.EqualError(t, err, "error selecting oauth2 blacklisted JTI with signature '50d858e0985ecc7f60418aaf0cc5ab587f42c2570a884095a9e8ccacd0f6545c': sql: no rows in result set")
+	assert.Nil(t, bjti)
+}
```

**Comment by alexlovelltroy** [comment]
Thanks for all your attention here!  

I'll have time to get back to a computer and test again tomorrow.

**Comment by james-d-elliott** [comment]
I've confirmed this and also found another bug (3 total including the log issue). I've fixed the other two discovered, not found the cause for this one yet.

**Comment by james-d-elliott** [comment]
Give `debug-oidc` a go, that branch doesn't seem to have the issue for me but I'm not entirely sure how it was fixed (or if it was fixed).

**Comment by alexlovelltroy** [comment]
I can confirm that my manual tests with the debug-oidc branch are now successful.  The only configuration changes I made to the oidc suite configuration were to add another client.

```yaml
identity_providers:
  oidc:
    clients:
      - client_id: 'oidc-tester-app-2'
        public: false
        require_pkce: true
        pkce_challenge_method: 'S256'
        redirect_uris:
          - 'https://oidc.example.com:8080/oauth2/callback'
        scopes:
          - 'offline_access'
          - 'offline'
          - 'authelia.bearer.authz'
          
        audience:
          - 'https://app1.example.com'
          - 'https://app2.example.com'
        grant_types:
          - 'authorization_code'
          - 'refresh_token'
        response_types:
          - 'code'
        response_modes:
          - 'form_post'
        consent_mode: 'explicit'
        require_pushed_authorization_requests: true
        # https://www.authelia.com/configuration/identity-providers/openid-connect/clients/#token_endpoint_auth_method
        # client_secret_basic, client_secret_post, client_secret_jwt, private_key_jwt, or none
        token_endpoint_auth_method: 'private_key_jwt'
        token_endpoint_auth_signing_alg: 'RS256'
        access_token_signed_response_key_id: 'd45a5e-rs256'
        jwks:
          - key_id: 'oidc-tester-app-2-key'
            algorithm: 'RS256'
            use: 'sig'
            key: |
              -----BEGIN PUBLIC KEY-----
             <snip>
              -----END PUBLIC KEY-----  
```

I created a simple python client for testing that I have shared as a gist.

https://gist.github.com/alexlovelltroy/ee5d5f3a034301e6f6ff225d39991462

**Comment by alexlovelltroy** [comment]
I have also now confirmed that my setup works with the :debug-oidc container.  Thanks!

**Comment by james-d-elliott** [comment]
Can you reconfirm the issue exists on v4.38.10 because I don't know that I did anything to fix it? 

Also it should now accept the issuer URL, token URL, or PAR URL as an audience.

**Comment by alexlovelltroy** [comment]
I can confirm that the behavior is different between the two versions.

##  authelia:debug-oidc

### Sanitized Client Logs
```
./client ./ochami auth  --client-id=oidc-tester-app-2 --auth-url=https://authelia.<snip> --redirect-uri=https://federation.<snip>/callback --insecure
Generated JWT client_assertion: eyJhbGciOiJSUzI1NiIsImtpZCI6Im9pZGMtdGVzdGVyLWFwcC0yLWtleSIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJvaWRjLXRlc3Rlci1hcHAtMiIsInN1YiI6Im9pZGMtdGVzdGVyLWFwcC0yIiwiYXVkIjoiaHR0cHM6Ly9hdXRoZWxpYS5vcGVuY2hhbWkuY29tL2FwaS9vaWRjL3Rva2VuIiwiaWF0IjoxNzI1NDQ4MjE3LCJleHAiOjE3MjU0NDgyNzcsImp0aSI6IjVlYmMwZjRiLTA5ZDUtNDlhNS04N2I1LWEyYzFmNDJlZWNjYSIsImNsaWVudF9pZCI6Im9pZGMtdGVzdGVyLWFwcC0yIn0.gn6ttAxW9pl5-v9EE7UBlsTNipcHQu160Nd2BsEInNjsEZYJo_2l_mc5g-NUylZcjm3RN_WTtfeWHzAk19jTds0XRBfmrX9jEFbIFNxvtCaGEl0PvGPJbB5HR2ZfSsesQiFjNHQI9CsrsBdZCyc5rQzPYF9PSqPDQvYsPIbMgp3TrTCb8LP5IVE1ING1x-vU63zqMJacAd7eBEP2kwyuQWv7K3mOgfj7Tl4gPq5E5oG_xa7-TuOClumwq43WZClmaUtqU7eaWc5ikg_fc3_73AimMeF7-CZi8lG8nl79Y3v5K4NVVEphOWCysyxgjaomzbN2m_YV-JapSMZz6DAk0A
Error in Pushed Authorization Request: 401 Client Error: Unauthorized for url: https://authelia.<snip>/api/oidc/pushed-authorization-request
```

### Sanitized Server Logs
```
2024-09-04 07:11:16 time="2024-09-04T21:11:16+10:00" level=debug msg="SQL No Rows Error returned (storage)"
2024-09-04 07:11:16 time="2024-09-04T21:11:16+10:00" level=debug msg="SQL No Rows Error returned (oidc)"
```

## 4.38.10

### Sanitized Client Logs
```
./client auth  --client-id=oidc-tester-app-2 --auth-url=https://authelia.<snip> --redirect-uri=https://federation.<snip>/callback --insecure
Generated JWT client_assertion: eyJhbGciOiJSUzI1NiIsImtpZCI6Im9pZGMtdGVzdGVyLWFwcC0yLWtleSIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJvaWRjLXRlc3Rlci1hcHAtMiIsInN1YiI6Im9pZGMtdGVzdGVyLWFwcC0yIiwiYXVkIjoiaHR0cHM6Ly9hdXRoZWxpYS5vcGVuY2hhbWkuY29tL2FwaS9vaWRjL3Rva2VuIiwiaWF0IjoxNzI1NDQ4NzIwLCJleHAiOjE3MjU0NDg3ODAsImp0aSI6IjkzYWFlZTdlLTM5ZWItNGE4MS1iMzc3LWVlODNhYWI4ZTdlZSIsImNsaWVudF9pZCI6Im9pZGMtdGVzdGVyLWFwcC0yIn0.UBuL-Tao7DLCFJAMjfm7fab9c2u00TXAaexwT6dgX8mokc2MBJYx-Ryh2oB-DXfqmNKD_xBfOlYzrgi2fF5ZmoU-G_KzO1rqVoWs4RbNDTKANtsq9trfitYO87J1XyJwfmRHQpBZWD1XALLkNHu5q2AIYa1Fti7YCnPTFedSDQEPygBCnQWtU_AWGKl7NacFT7-nBGIVJa05rWu4koA8aZAbnSMYkn2pelPkVfPQbbWwvhAEVOfJ7F0mAxwIHSExwOIWyQCCbWN3J44Y3skQb-xTb8pnXPKmL9xctYsG070zSg3zKY2cLRApf3aEJC85of8xCvnhNqIB5vXiEDZ-sQ
Error in Pushed Authorization Request: 401 Client Error: Unauthorized for url: https://authelia.<snip>/api/oidc/pushed-authorization-request
```

### Sanitized Server Logs
```
2024-09-04 07:18:40 time="2024-09-04T21:18:40+10:00" level=error msg="Pushed Authorization Request failed with error: Client authentication failed (e.g., unknown client, no client authentication included, or unsupported authentication method). The requested OAuth 2.0 Client could not be authenticated. The jti was already used. Claim 'jti' from 'client_assertion' MUST only be used once. error selecting oauth2 blacklisted JTI with signature '': sql: no rows in result set" method=POST path=/api/oidc/pushed-authorization-request remote_ip=192.168.65.1 stack="github.com/authelia/authelia/v4/internal/handlers/handler_oidc_authorization.go:176          OpenIDConnectPushedAuthorizationRequest\ngithub.com/authelia/authelia/v4/internal/middlewares/http_to_authelia_handler_adaptor.go:114 handleRouter.NewHTTPToAutheliaHandlerAdaptor.func23\ngithub.com/authelia/authelia/v4/internal/middlewares/bridge.go:54                            handleRouter.(*BridgeBuilder).Build.func16.1\ngithub.com/authelia/authelia/v4/internal/middlewares/headers.go:91                           SecurityHeadersNoStore.func1\ngithub.com/authelia/authelia/v4/internal/middlewares/headers.go:75                           SecurityHeadersCSPNoneOpenIDConnect.func1\ngithub.com/authelia/authelia/v4/internal/middlewares/headers.go:30                           SecurityHeadersBase.func1\ngithub.com/authelia/authelia/v4/internal/middlewares/cors.go:216                             handleRouter.(*CORSPolicy).Middleware.func24\ngithub.com/fasthttp/router@v1.5.2/router.go:441                                              (*Router).Handler\ngithub.com/authelia/authelia/v4/internal/middlewares/log_request.go:14                       handleRouter.LogRequest.func40\ngithub.com/authelia/authelia/v4/internal/middlewares/errors.go:38                            RecoverPanic.func1\ngithub.com/valyala/fasthttp@v1.55.0/server.go:2379                                           (*Server).serveConn\ngithub.com/valyala/fasthttp@v1.55.0/workerpool.go:224                                        (*workerPool).workerFunc\ngithub.com/valyala/fasthttp@v1.55.0/workerpool.go:196                                        (*workerPool).getCh.func1\nruntime/asm_arm64.s:1222                                                                     goexit"
```

**Comment by james-d-elliott** [comment]
Thanks a lot, fixed in 561d34fb47d9b73f6caa478b2482a1e0243eb1ae by a linter. The work in implementing them is paying for itself.
